File: backend/governance/src/test/java/com/commonsengine/governance/service/GovernanceServiceTest.java
package com.commonsengine.governance.service;

import com.commonsengine.governance.domain.Model.Proposal;
import com.commonsengine.governance.domain.Model.ProposalId;
import com.commonsengine.governance.domain.Model.ProposalStatus;
import com.commonsengine.governance.domain.Model.ProposalType;
import com.commonsengine.governance.domain.Model.StakeholderType;
import com.commonsengine.governance.domain.Model.Vote;
import com.commonsengine.governance.domain.Model.VoteChoice;
import com.commonsengine.governance.domain.Model.VoteResult;
import com.commonsengine.governance.infrastructure.persistence.GovernancePersistence.ProposalEntity;
import com.commonsengine.governance.infrastructure.persistence.GovernanceRepositories.ProposalRepository;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * 治理服务测试——从 Kotlin GovernanceServiceTest 转换。
 *
 * <p>适配 Java GovernanceService API：
 * <ul>
 *   <li>{@code createProposal(title, description, type, proposerId)}（默认讨论期 72h、投票期 48h）</li>
 *   <li>{@code openVoting(id)}——讨论期满后开启投票</li>
 *   <li>{@code castVote(proposalId, voterId, stakeholderType, choice, reason)}</li>
 *   <li>{@code tallyAndResolve(id)}——加权计票并决议</li>
 * </ul>
 *
 * <p>权重：WORKER 0.40 / CONSUMER 0.30 / COMMUNITY 0.30；得分 > 0 即通过。
 */
@SpringBootTest
@ActiveProfiles("test")
@Transactional
class GovernanceServiceTest {

    @Autowired
    private GovernanceService service;

    @Autowired
    private ProposalRepository proposalRepository;

    @Test
    void createProposalSetsDefaultDiscussionPeriod() {
        Proposal proposal = service.createProposal("调整分账比例", "描述",
                ProposalType.FEE_ADJUSTMENT, "member-1");
        long daysUntilVotingOpens = ChronoUnit.DAYS.between(
                Instant.now(), proposal.votingOpensAt());
        // 默认讨论期 72h ≈ 3 天
        assertTrue(daysUntilVotingOpens >= 2 && daysUntilVotingOpens <= 3,
                "讨论期应约 3 天（72h），实际: " + daysUntilVotingOpens);
    }

    @Test
    void createProposalSetsVotingClosesAfterVotingPeriod() {
        Proposal proposal = service.createProposal("修改章程", "描述",
                ProposalType.PLATFORM_RULE, "member-1");
        long hoursBetween = ChronoUnit.HOURS.between(
                proposal.votingOpensAt(), proposal.votingClosesAt());
        // 投票期默认 48h
        assertEquals(48, hoursBetween, "投票期应为 48 小时");
    }

    @Test
    void cannotStartVoteBeforeDiscussionDeadline() {
        Proposal proposal = service.createProposal("测试", "描述",
                ProposalType.POLICY_CHANGE, "member-1");
        // 讨论期未满，开启投票应失败
        assertThrows(Exception.class, () -> service.openVoting(proposal.id()));
    }

    @Test
    void weightedVotingWorksCorrectly() {
        Proposal proposal = service.createProposal("测试投票", "描述",
                ProposalType.POLICY_CHANGE, "member-1");

        // 直接将讨论截止时间设为过去，然后开启投票
        forceDiscussionOver(proposal.id());

        service.openVoting(proposal.id());

        // 投票：2 个劳动者赞成（+0.4 +0.4），1 个消费者反对（−0.3）
        service.castVote(proposal.id(), "w1", StakeholderType.WORKER, VoteChoice.APPROVE, null);
        service.castVote(proposal.id(), "w2", StakeholderType.WORKER, VoteChoice.APPROVE, null);
        service.castVote(proposal.id(), "c1", StakeholderType.CONSUMER, VoteChoice.REJECT, null);

        // 推进投票期到结束后再计票
        forceVotingOver(proposal.id());

        VoteResult result = service.tallyAndResolve(proposal.id());

        // +0.8 − 0.3 = +0.5 > 0 → 通过
        assertTrue(result.passed(), "赞成权重 0.8 > 反对 0.3，应通过");
        assertEquals(0.5, result.weightedScore(), 0.001);
    }

    @Test
    void onePersonOneVote() {
        Proposal proposal = service.createProposal("测试", "描述",
                ProposalType.POLICY_CHANGE, "member-1");

        forceDiscussionOver(proposal.id());
        service.openVoting(proposal.id());

        service.castVote(proposal.id(), "voter-1", StakeholderType.WORKER,
                VoteChoice.APPROVE, null);
        // 同一投票者再次投票应被拒绝
        assertThrows(Exception.class, () ->
                service.castVote(proposal.id(), "voter-1", StakeholderType.WORKER,
                        VoteChoice.REJECT, null));
    }

    @Test
    void weightedScoreRejectsWhenOppositionOutweighs() {
        Proposal proposal = service.createProposal("否决议案", "描述",
                ProposalType.POLICY_CHANGE, "member-1");

        forceDiscussionOver(proposal.id());
        service.openVoting(proposal.id());

        // 只有劳动者反对（−0.4），无人赞成 → 得分 −0.4 < 0 → 否决
        service.castVote(proposal.id(), "w1", StakeholderType.WORKER,
                VoteChoice.REJECT, null);

        forceVotingOver(proposal.id());
        VoteResult result = service.tallyAndResolve(proposal.id());

        assertFalse(result.passed(), "反对权重 0.4 > 0，应否决");
    }

    // ── 持久化测试 ─────────────────────────────────────

    @Test
    void createProposalPersistsAndCanBeFound() {
        Proposal proposal = service.createProposal("持久化测试", "测试描述",
                ProposalType.FEATURE_REQUEST, "member-persist");

        Proposal found = service.findById(proposal.id());
        assertNotNull(found);
        assertEquals("持久化测试", found.title());
        assertEquals("测试描述", found.description());
        assertEquals("member-persist", found.proposerId());
        assertEquals(ProposalStatus.DISCUSSION, found.status());
    }

    @Test
    void findByStatusReturnsPersistedList() {
        service.createProposal("提案A", "描述A", ProposalType.POLICY_CHANGE, "member-1");
        service.createProposal("提案B", "描述B", ProposalType.FEE_ADJUSTMENT, "member-2");

        List<Proposal> discussion = service.findByStatus(ProposalStatus.DISCUSSION);
        assertTrue(discussion.size() >= 2);
        assertTrue(discussion.stream().anyMatch(p -> p.title().equals("提案A")));
        assertTrue(discussion.stream().anyMatch(p -> p.title().equals("提案B")));
    }

    @Test
    void tallyVotesUpdatesProposalStatusToPassed() {
        Proposal proposal = service.createProposal("通过测试", "描述",
                ProposalType.POLICY_CHANGE, "member-1");

        forceDiscussionOver(proposal.id());
        service.openVoting(proposal.id());

        // 劳动者赞成（+0.4）+ 消费者赞成（+0.3）= +0.7 > 0
        service.castVote(proposal.id(), "w1", StakeholderType.WORKER,
                VoteChoice.APPROVE, null);
        service.castVote(proposal.id(), "c1", StakeholderType.CONSUMER,
                VoteChoice.APPROVE, null);

        forceVotingOver(proposal.id());
        service.tallyAndResolve(proposal.id());

        Proposal found = service.findById(proposal.id());
        assertEquals(ProposalStatus.PASSED, found.status());
    }

    @Test
    void tallyVotesUpdatesProposalStatusToRejected() {
        Proposal proposal = service.createProposal("否决测试", "描述",
                ProposalType.POLICY_CHANGE, "member-1");

        forceDiscussionOver(proposal.id());
        service.openVoting(proposal.id());

        // 只有劳动者反对（−0.4），无人赞成
        service.castVote(proposal.id(), "w1", StakeholderType.WORKER,
                VoteChoice.REJECT, null);

        forceVotingOver(proposal.id());
        service.tallyAndResolve(proposal.id());

        Proposal found = service.findById(proposal.id());
        assertEquals(ProposalStatus.REJECTED, found.status());
    }

    @Test
    void openVotingTransitionsStatusToVoting() {
        Proposal proposal = service.createProposal("投票状态测试", "描述",
                ProposalType.POLICY_CHANGE, "member-1");

        forceDiscussionOver(proposal.id());
        Proposal updated = service.openVoting(proposal.id());

        assertNotNull(updated);
        assertEquals(ProposalStatus.VOTING, updated.status());
    }

    @Test
    void proposalTypePersistsCorrectly() {
        Proposal proposal = service.createProposal("预算提案", "描述",
                ProposalType.FEE_ADJUSTMENT, "member-1");

        Proposal found = service.findById(proposal.id());
        assertEquals(ProposalType.FEE_ADJUSTMENT, found.type());
    }

    @Test
    void findVotesReturnsCastVotes() {
        Proposal proposal = service.createProposal("查票测试", "描述",
                ProposalType.POLICY_CHANGE, "member-1");

        forceDiscussionOver(proposal.id());
        service.openVoting(proposal.id());
        service.castVote(proposal.id(), "w1", StakeholderType.WORKER,
                VoteChoice.APPROVE, "支持");

        List<Vote> votes = service.findVotes(proposal.id());
        assertEquals(1, votes.size());
        assertEquals("w1", votes.get(0).voterId());
    }

    // ── 辅助方法 ───────────────────────────────────────

    /**
     * 直接通过 repository 将提案的投票开放时间设为过去，使讨论期「已满」。
     */
    private void forceDiscussionOver(ProposalId id) {
        ProposalEntity entity = proposalRepository.findById(id.value()).orElseThrow();
        entity.setVotingOpensAt(Instant.now().minusSeconds(1));
        proposalRepository.save(entity);
    }

    /**
     * 直接通过 repository 将提案的投票关闭时间设为过去，使投票期「已结束」。
     */
    private void forceVotingOver(ProposalId id) {
        ProposalEntity entity = proposalRepository.findById(id.value()).orElseThrow();
        entity.setVotingClosesAt(Instant.now().minusSeconds(1));
        proposalRepository.save(entity);
    }
}

File: backend/governance/src/test/kotlin/com/commonsengine/governance/service/GovernanceServiceTest.kt
package com.commonsengine.governance.service

import com.commonsengine.governance.domain.ProposalStatus
import com.commonsengine.governance.domain.ProposalType
import com.commonsengine.governance.domain.StakeholderType
import com.commonsengine.governance.domain.VoteChoice
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.time.temporal.ChronoUnit

@SpringBootTest
@ActiveProfiles("test")
@Transactional
class GovernanceServiceTest {

    @Autowired
    private lateinit var service: GovernanceService

    @Test
    fun `create proposal sets 30 day discussion for regular proposals`() {
        val proposal = service.createProposal("调整分账比例", "描述", "member-1", ProposalType.SETTLEMENT_RULE)
        val daysUntilDeadline = ChronoUnit.DAYS.between(Instant.now(), proposal.discussionDeadline)
        assertTrue(daysUntilDeadline in 29..30, "讨论期应约 30 天，实际: $daysUntilDeadline")
    }

    @Test
    fun `create proposal sets 45 day discussion for charter amendments`() {
        val proposal = service.createProposal("修改章程", "描述", "member-1", ProposalType.CHARTER_AMENDMENT)
        val daysUntilDeadline = ChronoUnit.DAYS.between(Instant.now(), proposal.discussionDeadline)
        assertTrue(daysUntilDeadline in 44..45, "章程修改讨论期应约 45 天，实际: $daysUntilDeadline")
    }

    @Test
    fun `cannot start vote before discussion deadline`() {
        val proposal = service.createProposal("测试", "描述", "member-1")
        assertThrows<IllegalArgumentException> { service.startVote(proposal.id) }
    }

    @Test
    fun `weighted voting works correctly`() {
        // Create proposal with past deadline directly via repository
        val proposal = service.createProposal("测试投票", "描述", "member-1")
        // Use the service's internal method to set past deadline by creating a proposal
        // then starting vote after deadline
        // Since we can't directly modify the entity, we test with a proposal
        // whose deadline is manually set via the repository

        // For this test, we need to bypass the discussion deadline check.
        // We'll use the repository directly to set a past deadline.
        val proposalRepo = service.javaClass.getDeclaredField("proposalRepository")
        proposalRepo.isAccessible = true
        @Suppress("UNCHECKED_CAST")
        val repo = proposalRepo.get(service) as com.commonsengine.governance.infrastructure.persistence.ProposalRepository

        val entity = repo.findByProposalId(proposal.id.value)!!
        entity.discussionDeadline = Instant.now().minus(1, ChronoUnit.DAYS)
        repo.save(entity)

        service.startVote(proposal.id)

        // 投票：2个劳动者赞成（0.8），1个消费者反对（0.3）
        service.castVote(proposal.id, "w1", StakeholderType.WORKER, VoteChoice.YES)
        service.castVote(proposal.id, "w2", StakeholderType.WORKER, VoteChoice.YES)
        service.castVote(proposal.id, "c1", StakeholderType.CONSUMER, VoteChoice.NO)

        val result = service.tallyVotes(proposal.id)

        assertTrue(result.passed, "赞成 0.8 > 反对 0.3，应通过")
        assertEquals(0.8, result.yesWeighted, 0.001)
        assertEquals(0.3, result.noWeighted, 0.001)
    }

    @Test
    fun `one person one vote`() {
        val proposal = service.createProposal("测试", "描述", "member-1")

        // Set past deadline and start voting
        val proposalRepo = service.javaClass.getDeclaredField("proposalRepository")
        proposalRepo.isAccessible = true
        @Suppress("UNCHECKED_CAST")
        val repo = proposalRepo.get(service) as com.commonsengine.governance.infrastructure.persistence.ProposalRepository

        val entity = repo.findByProposalId(proposal.id.value)!!
        entity.discussionDeadline = Instant.now().minusSeconds(1)
        entity.status = ProposalStatus.VOTING
        repo.save(entity)

        service.castVote(proposal.id, "voter-1", StakeholderType.WORKER, VoteChoice.YES)
        assertThrows<IllegalArgumentException> {
            service.castVote(proposal.id, "voter-1", StakeholderType.WORKER, VoteChoice.NO)
        }
    }

    @Test
    fun `charter amendment requires two thirds majority`() {
        val proposal = service.createProposal("修宪", "描述", "member-1", ProposalType.CHARTER_AMENDMENT)

        val proposalRepo = service.javaClass.getDeclaredField("proposalRepository")
        proposalRepo.isAccessible = true
        @Suppress("UNCHECKED_CAST")
        val repo = proposalRepo.get(service) as com.commonsengine.governance.infrastructure.persistence.ProposalRepository

        val entity = repo.findByProposalId(proposal.id.value)!!
        entity.discussionDeadline = Instant.now().minusSeconds(1)
        entity.status = ProposalStatus.VOTING
        repo.save(entity)

        // 赞成 60%（简单多数但不到 2/3）
        service.castVote(proposal.id, "c1", StakeholderType.CONSUMER, VoteChoice.YES)  // 0.3
        service.castVote(proposal.id, "cm1", StakeholderType.COMMUNITY, VoteChoice.YES) // 0.3
        service.castVote(proposal.id, "w1", StakeholderType.WORKER, VoteChoice.NO)    // 0.4

        val result = service.tallyVotes(proposal.id)
        assertFalse(result.passed, "修宪需 2/3 多数，60% 不够")
    }

    // ── 持久化测试 ─────────────────────────────────────

    @Test
    fun `create proposal persists and can be found`() {
        val proposal = service.createProposal("持久化测试", "测试描述", "member-persist")

        val found = service.findProposal(proposal.id)
        assertNotNull(found)
        assertEquals("持久化测试", found!!.title)
        assertEquals("测试描述", found.description)
        assertEquals("member-persist", found.proposedBy)
        assertEquals(ProposalStatus.DISCUSSION, found.status)
    }

    @Test
    fun `find non-existent proposal returns null`() {
        val found = service.findProposal(com.commonsengine.governance.domain.ProposalId("no-such-proposal"))
        assertNull(found)
    }

    @Test
    fun `find all proposals returns persisted list`() {
        service.createProposal("提案A", "描述A", "member-1")
        service.createProposal("提案B", "描述B", "member-2")

        val all = service.findAllProposals()
        assertTrue(all.size >= 2)
        assertTrue(all.any { it.title == "提案A" })
        assertTrue(all.any { it.title == "提案B" })
    }

    @Test
    fun `tally votes updates proposal status to approved`() {
        val proposal = service.createProposal("通过测试", "描述", "member-1")

        val proposalRepo = service.javaClass.getDeclaredField("proposalRepository")
        proposalRepo.isAccessible = true
        @Suppress("UNCHECKED_CAST")
        val repo = proposalRepo.get(service) as com.commonsengine.governance.infrastructure.persistence.ProposalRepository

        val entity = repo.findByProposalId(proposal.id.value)!!
        entity.discussionDeadline = Instant.now().minusSeconds(1)
        entity.status = ProposalStatus.VOTING
        repo.save(entity)

        // 劳动者赞成（0.4）+ 消费者赞成（0.3）= 0.7 > 0.5
        service.castVote(proposal.id, "w1", StakeholderType.WORKER, VoteChoice.YES)
        service.castVote(proposal.id, "c1", StakeholderType.CONSUMER, VoteChoice.YES)

        service.tallyVotes(proposal.id)

        val found = service.findProposal(proposal.id)!!
        assertEquals(ProposalStatus.APPROVED, found.status)
    }

    @Test
    fun `tally votes updates proposal status to rejected`() {
        val proposal = service.createProposal("否决测试", "描述", "member-1")

        val proposalRepo = service.javaClass.getDeclaredField("proposalRepository")
        proposalRepo.isAccessible = true
        @Suppress("UNCHECKED_CAST")
        val repo = proposalRepo.get(service) as com.commonsengine.governance.infrastructure.persistence.ProposalRepository

        val entity = repo.findByProposalId(proposal.id.value)!!
        entity.discussionDeadline = Instant.now().minusSeconds(1)
        entity.status = ProposalStatus.VOTING
        repo.save(entity)

        // 只有劳动者反对（0.4），无人赞成 → 0% < 50%
        service.castVote(proposal.id, "w1", StakeholderType.WORKER, VoteChoice.NO)

        service.tallyVotes(proposal.id)

        val found = service.findProposal(proposal.id)!!
        assertEquals(ProposalStatus.REJECTED, found.status)
    }

    @Test
    fun `start vote transitions status to voting`() {
        val proposal = service.createProposal("投票状态测试", "描述", "member-1")

        val proposalRepo = service.javaClass.getDeclaredField("proposalRepository")
        proposalRepo.isAccessible = true
        @Suppress("UNCHECKED_CAST")
        val repo = proposalRepo.get(service) as com.commonsengine.governance.infrastructure.persistence.ProposalRepository

        val entity = repo.findByProposalId(proposal.id.value)!!
        entity.discussionDeadline = Instant.now().minus(1, ChronoUnit.DAYS)
        repo.save(entity)

        val updated = service.startVote(proposal.id)
        assertNotNull(updated)
        assertEquals(ProposalStatus.VOTING, updated!!.status)
    }

    @Test
    fun `proposal type persists correctly`() {
        val proposal = service.createProposal("预算提案", "描述", "member-1", ProposalType.BUDGET_ALLOCATION)

        val found = service.findProposal(proposal.id)!!
        assertEquals(ProposalType.BUDGET_ALLOCATION, found.type)
    }
}
